File: 1/part2.rb
# frozen_string_literal: true

numbers = File.read(
  File.join(
    File.dirname(__FILE__),
    './input.txt'
  )
).split.map(&:to_i)

previous_compute = numbers[0..2].sum

result = numbers[2..].each_with_index.reduce(0) do |count, (_, i)|
  return count if numbers[i + 2].nil?

  current_compute = numbers[i..i + 2].sum

  r = if current_compute > previous_compute
        count + 1
      else
        count
      end

  previous_compute = current_compute

  r
end

print result

File: 2023/01/part2.exs
file = 'input.txt'
input = File.read!(Path.join(Path.dirname(__ENV__.file), file))
lines = String.split(input)

replacements = %{
  "one" => "one1one",
  "two" => "two2two",
  "three" => "three3three",
  "four" => "four4four",
  "five" => "five5five",
  "six" => "six6six",
  "seven" => "seven7seven",
  "eight" => "eight8eight",
  "nine" => "nine9nine",
}

convert_string_to_digit = fn line ->
  Enum.reduce(replacements, line, fn {word, digit}, acc ->
    String.replace(acc, word, digit)
  end)
end

first_and_last_digit_combined = fn line ->
  matches = List.flatten(Regex.scan(~r/[0-9]/, line))

  fe = hd(matches)
  le = hd(Enum.reverse(matches))
  concat = fe <> le
  String.to_integer(concat)
end

true_numbers = Enum.map(lines, convert_string_to_digit)
numbers = Enum.map(true_numbers, first_and_last_digit_combined)

IO.puts(numbers |> Enum.sum)
